Initial implementation

alecgibson opened the pull request:
This is an initial implementation of ShareDB's [`MilestoneDB`][1]. The
bare bones of `MongoMilestoneDB` connection logic are based on the
[`sharedb-mongo`][2] database adapter.

This adapter is a slight departure from other ShareDB code, as it is
written in ES6. This is because this code is server-side-only, so we
don't need to support legacy browsers and all LTS Node versions support
ES6. The move should ease maintenance and understanding of asynchronous
code through the use of `Promise`s, which the [`mongodb`][3] library
supports out of the box.

This library notably uses v2 of `mongodb` instead of the newer v3. This
is to stay consistent with `sharedb-mongo`, so in theory the same config
(or style of config) can be used in both `sharedb-mongo` and in
`sharedb-milestone-mongo`.

We also introduce usage of [`eslint`][4] instead of [`jshint`][5], which
is used in other projects. `eslint` supports a much, much wider range of
style rules and helps to maintain a uniform coding style.

## Dependencies

- [x] https://github.com/share/sharedb/pull/244

[1]: share/sharedb#236
[2]: share/sharedb-mongo
[3]: mongodb.github.io/node-mongodb-native
[4]: https://eslint.org/
[5]: http://jshint.com/

alecgibson commented:
I'm happy to discuss what language this should be written in. I've written it in ES6 as the most modern vanilla JS that LTS Node supports (ideally I'd have liked ES7 for `async`/`await`).

Other options include:

  - Keep it in ES5, which would force us back into callbacks. This would be more consistent with the rest of ShareDB, but it also forces us to do things like [basically reimplement a `Promise`][1], which we obviously get for free in ES6. I also think it's particularly weird to not use the `Promise`s returned directly by `mongodb`.
  - Use a transpiler, which would let us use ES7, or even move to TypeScript. I personally love TypeScript, but I realise some people think that this can raise the maintenance burden for people switching between libraries in the same org. I personally think a good IDE goes a long way to basically removing this burden, but I also switch between Vanilla and TypeScript on basically an hourly basis, so perhaps I'm more used to it than some. I can definitely see an argument for staying Vanilla on an open source library, especially since I've sometimes looked at eg CoffeeScript libraries and had no idea what's going on.

[1]: https://github.com/share/sharedb-mongo/blob/2d579ddb80781e987707076e932cd4e01ca066ef/index.js#L113-L119

alecgibson commented on lib/mongo-milestone-db.js:
```diff
@@ -0,0 +1,181 @@
+      mongo = options.mongo;
+    }
+
+    options = options || {};
```
NB: The motivation for deleting these options is because the Mongo client validates them. In fact, if you pass in the option `validateOptions`, then your mongo connection will actually fail if we don't tidy these up.

https://github.com/mongodb/node-mongodb-native/blob/bd4fb531a7f599bb6cf50ebbab3b986f191a7ef8/lib/mongo_client.js#L53-L57

I'll add a comment in the code for this, too.

alecgibson commented:
@ericyhwang I've addressed the issues we discussed in the meeting. Note that this PR now has a dependency on an update to the Driver API in ShareDB core: https://github.com/share/sharedb/pull/244

ericyhwang commented on lib/mongo-milestone-db.js:
```diff
@@ -0,0 +1,227 @@
+  static _shallowClone(object) {
+    if (typeof object !== 'object') return object;
+
+    const out = {};
```
This can be written as:
```
return Object.assign({}, object)
```

[`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) copies enumerable own properties from the second argument (then the third arg, fourth arg, etc. if present) onto the first object. Otherwise, if doing a manual loop, a `object.hasOwnProperty()` check would be needed to avoid copying over enumerable prototype properties on `object`.

alecgibson replied:
Ha! I somehow completely forgot that I was in ES6 🤦🏻‍♂️

ericyhwang commented on lib/mongo-milestone-db.js:
```diff
@@ -0,0 +1,227 @@
+          // collection this won't be a problem, but this is a dangerous mechanism.
+          // Perhaps we should only warn instead of creating the indexes, especially
+          // when there is a lot of data in the collection.
+          return collection.createIndex({id: 1, v: 1}, {background: true, unique: true});
```
I'm wondering, will it get confusing to have both `_id` and `id` properties on these Mongo documents?

One idea is to use `d`, to match what the Mongo ops collections use to indicate the Share doc id. I'm open to other ideas, or staying with `id` if we end up deciding that's fine.

alecgibson replied:
I mostly like using `id` directly because it avoids any need for mapping the snapshots when reading/writing, but if you feel strongly we can change it. I think it quickly becomes clear which `id` is which when you poke around in the data at all (if you are so inclined). And if you're poking around, you're hopefully both:

  - familiar with Mongo, and therefore unsurprised by an `_id` field containing an `ObjectID`
  - familiar with ShareDB, and therefore unsurprised by an `id` field corresponding to the ShareDB doc ID

What do you think?

ericyhwang replied:
From Share PR meeting:
* Nate prefers to use `d` as the name
* See if we can use projection to rename it to `id` when reading

alecgibson commented:
@ericyhwang I've addressed the `Object.assign` thing. I guess we just need to discuss the `id` vs `_id` thing. I'm personally happy to leave it, but if you feel strongly, I can map `id` to `d`.

alecgibson commented:
@ericyhwang I've bumped the ShareDB version so this no longer points at a fork, and I've added the `id` -> `d` mapping we discussed.

Unfortunately, I don't think the Mongo lets you use `string` values for renaming projections (there's a bunch of code [like this][1] floating around that just changes values to `1`), so I've just added a couple of methods for mapping to and from the database representation.

[1]: https://github.com/mongodb/node-mongodb-native/blob/2.2/lib/collection.js#L217

alecgibson commented:
Released as v0.1.0. @ericyhwang and @nateps I've added you both as owners of the package.
